portal-impl/src/com/liferay/portlet/internal/PortletURLImpl.java

```diff
@@ -728,7 +760,18 @@ public void setWindowStateRestoreCurrentView(
 
 	@Override
 	public String toString() {
-		if (_toString != null) {
+		LiferayMutablePortletParameters mutableActionParameters =
+			(LiferayMutablePortletParameters)_mutableActionParameters;
+
+		LiferayMutablePortletParameters mutableResourceParameters =
+			(LiferayMutablePortletParameters)_mutableResourceParameters;
+
+		if (!_mutableRenderParameters.isMutated() &&
```

Comment:
@shuyangzhou Here you can see an optimization that I introduced that will only re-generate the toString value if the mutable action/render/resource parameters have actually mutated. I know it doesn't look pretty, but it works. If you can make it look more pretty I'd be interested to see what you come up with.